### PR TITLE
Use external prices for conversion of fees in autopilot 2

### PR DESCRIPTION
**Pull request by sunce86**
# Description
Same as https://github.com/cowprotocol/services/pull/2863 but now updates the old code that is currently in use, so that it is aligned with how new code is converting fees to sell token.

After this change is merged, no fee mismatches should be detected by our recently merged automatic checker: 

https://github.com/cowprotocol/services/pull/2865/files#diff-1b61656a519f2e07d2dd2edbb8ce74fff0101b3fadb8ee6af192da88cd89d065R211

https://aws-es.cow.fi/_dashboards/goto/5f1e20d61c056e15308d7aed7d2af2f1?security_tenant=global

The entire file that is touched by this PR will be deleted in https://github.com/cowprotocol/services/pull/2861 but anyway I find it useful to merge this PR for testing purposes and making sure no refactoring bugs are introduced!

## How to test
Existing unit tests

### CODE REVIEW

File: crates/autopilot/src/decoded_settlement.rs [hunk 1]
```diff
@@ -1372,19 +1390,19 @@ mod tests {
...
             ),
             (
                 addr!("da816459f1ab5631232fe5e97a05bbbb94970c95"),
-                U256::from(1000000000000000000u128),
+                U256::from(490332962282438u128),
             ),
             (
                 addr!("fbeb78a723b8087fd2ea7ef1afec93d35e8bed42"),
-                U256::from(1000000000000000000u128),
+                U256::from(2812973786294373u128),
```
Comment (squadgazzz) [inline crates/autopilot/src/decoded_settlement.rs:1405]: How these values were calculated?

Comment (sunce86) [reply crates/autopilot/src/decoded_settlement.rs:1405]: These are external prices taken from solver competition entry for this tx

**Review by fleupold — APPROVED** [review]

File: crates/autopilot/src/decoded_settlement.rs [hunk 2]
```diff
@@ -355,35 +356,49 @@ impl DecodedSettlement {
...
 
         Some(Fees {
             order: trade.order_uid,
             kind,
-            sell: fee,
+            sell: match trade.flags.order_kind() {
+                OrderKind::Buy => fee,
+                OrderKind::Sell => big_rational_to_u256(
+                    &(u256_to_big_rational(&fee) * external_prices.price(buy_token)?
+                        / external_prices.price(sell_token)?),
```
Comment (MartinquaXD) [inline crates/autopilot/src/decoded_settlement.rs:398]: I'd say `checked_div` just to be safe

File: crates/autopilot/src/decoded_settlement.rs [hunk 3]
```diff
@@ -331,8 +331,8 @@ impl DecodedSettlement {
         let sell_token = self.tokens.get(sell_index)?;
         let buy_token = self.tokens.get(buy_index)?;
 
-        let (kind, fee) = match trade.fee_amount.is_zero() {
-            false => (FeeKind::User, trade.fee_amount),
+        let (kind, fee, surplus_token) = match trade.fee_amount.is_zero() {
```
Comment (MartinquaXD) [inline crates/autopilot/src/decoded_settlement.rs:334]: Might be slightly nicer to read the code if the surplus token gets decided in a separate match statement:
```rust
        let surplus_token = match (trade.fee_amount.is_zero(), trade.flags.order_kind()) {
            (false, _) => sell_token, // pre-signed fee is always in the sell token
            (_, OrderKind::Buy) => sell_token,
            (_, OrderKind::Sell) => buy_token,
        };
```

**Review by squadgazzz — APPROVED** [review]
LGTM

File: crates/autopilot/src/decoded_settlement.rs (hunk 2 above)
Comment (sunce86) [reply crates/autopilot/src/decoded_settlement.rs]: makes sense 👍

**Review by MartinquaXD — APPROVED** [review]

**Review by Mateo-mro — APPROVED** [review]

**Comment by sunce86** [comment]
Closing because of https://github.com/cowprotocol/services/pull/2891

**Comment by sunce86** [comment]
Reopen as after discussion it's decided to use external prices for conversion of fees from surplus token to sell token.

File: crates/e2e/tests/e2e/protocol_fee.rs [hunk 4]
```diff
@@ -326,36 +325,6 @@ async fn combined_protocol_fees(web3: Web3) {
...
-        market_order_token_balance
-    );
-    assert_approximately_eq!(
-        limit_executed_surplus_fee_in_buy_token,
-        limit_order_token_balance
-    );
-    assert_approximately_eq!(
-        partner_fee_executed_surplus_fee_in_buy_token,
-        partner_fee_order_token_balance
-    );
```
Comment (sunce86) [inline crates/e2e/tests/e2e/protocol_fee.rs:358]: Executed surplus fee cant longer be used to check the amount of fee kept in the settlement contract.

**Comment by sunce86** [comment]
Closing as https://github.com/cowprotocol/services/pull/2891 is merged.

**Pull request closed without merging**